Retrieve images using MusicBrianz ID

jcass77 opened the pull request:
This represents an effort to make sure that requests to last.fm for album covers and artist images will always be unique:

- Now uses unique MusicBrainz ID whenever it is available. This should be more reliable than simply relying on album and artist names.

- `getCoverFromLastFm` now also requires that *both* the album and artist names be supplied if the MusicBrainz ID is not available.

- In the absence of a MusicBrainzID, `getArtistImage` could potentially still retrieve the wrong artist image from last.fm (e.g. a different artist with the same or a similar name). I'm not sure if there is away around this as the [last.fm API](http://www.last.fm/api/show/artist.getInfo) does not provide for anything else at this stage.

jcass77 commented:
137593c586f2291c2ff88ca593d7e489e288fc80 fixes last.fm call signature.

kingosticks commented:
Oh nice. Now if I could just remember my test case where the wrong cover was requested.

kingosticks commented on mopidy_musicbox_webclient/static/js/images.js:
```diff
@@ -61,47 +64,92 @@
+            var musicBrainzID = ''
+
+            if (track && track.album && track.album.artists && track.album.artists.length > 0) {
                 // First look for the artist in the album
```
Out of scope for this PR. But does this make sense do you think? Seems that for various artists compilations this does the wrong thing and it should use `track.artists` first.

jcass77 replied:
This is easy enough to change and I agree using track artist before album artist will probably make sense for most situations.

kingosticks commented on mopidy_musicbox_webclient/static/js/images.js:
```diff
@@ -61,47 +64,92 @@
-                    }
+
+            var _this = this
+            mopidy.library.lookup({'uris': [uri]}).then(function (resultDict) {
```
Shouldn't this try a `mopidy.library.getImages` first? Am I missing something?

jcass77 replied:
I have it as `mopidy.library.getImages` just retrieving **album** cover art. We're looking for the generic **artist** image here and need the track as a starting point.

kingosticks replied:
Sure, I appreciate the different goals here. My point is that if you pass getImages an artist_uri then, depending on the backend, it may be able to provide you with an artist image. If not, you can fallback to using track metadata to do a last.fm request.

jcass77 replied:
Thanks Nick, I didn't realise that backends are supposed to implement `getImages` to cater for artist URIs as well (don't think I've ever seen that happen before).

It makes sense to try that first and fall back to the track metadata after.

kingosticks replied:
Mopidy-spotify should, it can get images for any resource I think. Mopidy-tunein handles either artists or albums (can't remember which, it's a bit arbitrary) as station and show logos. Others, not so sure.

kingosticks commented on mopidy_musicbox_webclient/static/js/images.js:
```diff
@@ -61,47 +64,92 @@
+                    params.artist = artist.name
+                }
+
+                // TODO: Artist name may not be unique, which could result in the wrong images being retrieved.
```
Turns out this is exactly the issue I was having. Since we don't use Mopidy's `getImages` it falls back to last.fm and then picks the cover art for the wrong artist.

jcass77 replied:
Yes. I'm not sure what to do about this except insist on using unique `MusicBrainz IDs` (which I think will not be available in most instances).

We can probably retrieve the album info from last.fm first, and then rely on the artist info that it sends back to make sure that we retrieve the correct image. It might require a bit of rework on the UI side, but I'll take a quick stab at it.

kingosticks replied:
How would you do that? Do a last.fm artist search (I assume you can, not checked) first and then try and pick the correct artist from multiple results, at which point you'll then have the musicbrainz Id to use for image requests? Would the UI be to help pick the correct artist or to just flag the image as wrong and try again? Both sound like a fair bit of work.

kingosticks commented on mopidy_musicbox_webclient/static/js/process_ws.js:
```diff
@@ -229,7 +229,7 @@ function processArtistResults (resultArr) {
     var artistname = getArtist(resultArr)
     $('#h_artistname, #artistpopupname').html(artistname)
-    getArtistImage(artistname, '#artistviewimage, #artistpopupimage', 'extralarge')
+    coverArt.getArtistImage(resultArr[0].uri, '#artistviewimage, #artistpopupimage', 'extralarge', mopidy)
```
I would have expected to pass the artist (rather than track) uri in here since we have it.

jcass77 replied:
This was an oversight on my part to start with, but we'll probably end up doing a lookup on the track at this point anyway so that we can ask last.fm for the track info first, and then retrieve the artist image using the artist's MusicBrainz ID that it provides.

jcass77 commented:
02cdb90ee8b7be3c9a7d55a09b832878c7f04936 is pretty much a complete rewrite, and addresses all of the comments.

The PR now also ensures that artist images will only be retrieved from last.fm using the unique MusicBrainz identifier (avoids having the wrong image being displayed if multiple artists with the same or similar names exist in the last.fm database).

jcass77 commented:
Any further comments / suggestions for this PR before we merge?

kingosticks commented:
It's working great but I've not had a chance to peruse the code quite yet, Will do so at lunchtime.

kingosticks commented on mopidy_musicbox_webclient/static/js/gui.js:
```diff
@@ -128,7 +128,7 @@ function setSongInfo (data) {
     if (data.track.album && data.track.album.name) {
         $('#modalalbum').html('<a href="#" onclick="return library.showAlbum(\'' + data.track.album.uri + '\');">' + data.track.album.name + '</a>')
-        coverArt.getCover(data.track.uri, '#infocover, #controlspopupimage', 'extralarge', mopidy)
+        images.setAlbumImage(data.track.uri, '#infocover, #controlspopupimage', mopidy)
```
We could move this out of the if statement and leave `setAlbumImage` to sort `'images/default_cover.png'` for us.

jcass77 replied:
Yes. Good catch!

kingosticks commented on mopidy_musicbox_webclient/static/js/process_ws.js:
```diff
@@ -229,7 +229,8 @@ function processArtistResults (resultArr) {
     var artistname = getArtist(resultArr)
     $('#h_artistname, #artistpopupname').html(artistname)
-    getArtistImage(artistname, '#artistviewimage, #artistpopupimage', 'extralarge')
+    var artistUri = resultArr[0].artists[0] ? resultArr[0].artists[0].uri : ''
```
Isn't `resultArr.uri` always the uri you want?

jcass77 replied:
You're right, I'll change it.

kingosticks commented on mopidy_musicbox_webclient/static/js/images.js:
```diff
@@ -4,105 +4,279 @@
+            var artist = images._getArtistInfo(track)
+            var artistPromise = $.Deferred()
+
+            if (!track || (!track.musicbrainz_id && !track.name) || (!track.musicbrainz_id && (!artist || !artist.name))) {
```
Isn't the mbid enough even without `track.name` or `artist.name` ?

jcass77 replied:
Yes. Isn't that what this if testing for? It is confusing because it is stated in the negative, but I read it as:

1. we need a track, and then
2. we need either the mbid, or the track name
3. if we do not have mbid, then we need the artist name as well

So we should then be able to do a lookup using either ``mbid``, or ``track.name`` **AND** ``artist.name``.

kingosticks replied:
Yeh you are right but just writing those conditions might be clearer:
```
if (! (track && (track.musicbrainz_id or (track.name and artist and artist.name))))
```

jcass77 replied:
Done. I also applied the same structure change to line 80.

kingosticks commented on mopidy_musicbox_webclient/static/js/images.js:
```diff
@@ -4,105 +4,279 @@
+        _setLastFmArtistImage: function (mbid, img_element, size) {
+            if (!mbid) {
+                // Avoid expensive last.fm call if tag information is missing.
+                $(img_element).attr('src', images.DEFAULT_ALBUM_URL)
```
Should this be `DEFAULT_ARTIST_URL` ?

jcass77 replied:
Absolutely. Isn't it impressive how I managed to get this mixed up, and then write a test that passes as well?

kingosticks commented:
The fact that this has tests is actually mind blowing. Really great work.

jcass77 commented:
Thanks for reviewing. I've addressed all of the comments in 3a7539aa02ce4eb5e2a47ef4460698e7a1f0cebe.

kingosticks commented:
LGTM. Squash and merge (that is a pretty handy new feature)?
